plan/statistics/statistics.go

```diff
@@ -311,19 +375,176 @@ func estimateNDV(sc *variable.StatementContext, count int64, samples []types.Dat
+			if !knowCount && bucketIdx+1 == bucketCount {
+				col.mergeBuckets(bucketIdx)
+				valuesPerBucket *= 2
+				bucketIdx = (bucketIdx + 1) / 2
+				if bucketIdx == 0 {
+					lastNumber = 0
+				} else {
+					lastNumber = col.Numbers[bucketIdx-1]
+				}
+			}
+			// We may merge buckets, so we should check it again.
+			if col.Numbers[bucketIdx]+1-lastNumber <= valuesPerBucket {
+				col.Numbers[bucketIdx]++
+				col.Values[bucketIdx] = data
+				col.Repeats[bucketIdx] = 0
+			} else {
+				lastNumber = col.Numbers[bucketIdx]
+				bucketIdx++
+				col.Numbers = append(col.Numbers, lastNumber+1)
+				col.Values = append(col.Values, data)
+				col.Repeats = append(col.Repeats, 0)
+			}
+			col.NDV++
+		}
+	}
+	if isPK {
+		t.Columns[offset] = col
+	} else {
+		t.Indices[offset] = col
+	}
+	return nil
+}
+
+func copyFromIndexColumns(ind *Column, id, numBuckets int64) (*Column, error) {
+	col := &Column{
+		ID:      id,
+		NDV:     ind.NDV,
+		Numbers: ind.Numbers,
+		Values:  make([]types.Datum, 0, numBuckets),
+		Repeats: ind.Repeats,
+	}
+	for _, val := range ind.Values {
+		if val.GetBytes() == nil {
+			break
+		}
+		data, err := codec.Decode(val.GetBytes(), 1)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		col.Values = append(col.Values, data[0])
+	}
+	return col, nil
+}
+
 // NewTable creates a table statistics.
-func NewTable(sc *variable.StatementContext, ti *model.TableInfo, ts, count, numBuckets int64, columnSamples [][]types.Datum) (*Table, error) {
+func NewTable(sc *variable.StatementContext, ti *model.TableInfo, ts, count, numBuckets int64, columnSamples [][]types.Datum,
+	colOffsets []int, indResults []ast.RecordSet, indOffsets []int, pkResult ast.RecordSet, pkOffset int) (*Table, error) {
+	if count == 0 {
+		return PseudoTable(ti), nil
+	}
 	t := &Table{
 		info:    ti,
 		TS:      ts,
 		Count:   count,
-		Columns: make([]*Column, len(columnSamples)),
+		Columns: make([]*Column, len(ti.Columns)),
+		Indices: make([]*Column, len(ti.Indices)),
+	}
+	for i, offset := range colOffsets {
+		err := t.buildColumn(sc, offset, columnSamples[i], numBuckets)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+	if pkOffset != -1 {
+		err := t.buildIndexColumn(sc, pkOffset, pkResult, numBuckets, true)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
 	}
-	for i, sample := range columnSamples {
-		err := t.buildColumn(sc, i, sample, defaultBucketCount)
+	for i, offset := range indOffsets {
+		err := t.buildIndexColumn(sc, offset, indResults[i], numBuckets, false)
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
+		if len(ti.Indices[offset].Columns) == 1 {
+			for j, col := range ti.Columns {
+				if col.Name.L == ti.Indices[offset].Columns[0].Name.L && t.Columns[j] == nil {
+					t.Columns[j], err = copyFromIndexColumns(t.Indices[offset], col.ID, numBuckets)
+					if err != nil {
+						return nil, errors.Trace(err)
+					}
+					break
+				}
+			}
+		}
+	}
+	if t.Count == 0 {
```

Comment:
There may be cases that we have no columnSamples, and only after we build the index columns that we can know it is 0, so I also checked it here. I will add comment for it.